pkg/services/ngalert/state/cache.go

```diff
@@ -44,13 +45,23 @@ func (c *cache) getOrCreate(ctx context.Context, log log.Logger, alertRule *ngMo
 	}
 	var states *ruleStates
 	if states, ok = orgStates[alertRule.UID]; !ok {
-		states = &ruleStates{states: make(map[string]*State)}
+		states = &ruleStates{states: make(map[string]*State), ruleVersion: alertRule.Version}
 		c.states[alertRule.OrgID][alertRule.UID] = states
 	}
 	return states.getOrCreate(ctx, log, alertRule, result, extraLabels, externalURL)
 }
 
 func (rs *ruleStates) getOrCreate(ctx context.Context, log log.Logger, alertRule *ngModels.AlertRule, result eval.Result, extraLabels data.Labels, externalURL *url.URL) *State {
+	if alertRule.Version > rs.ruleVersion {
```

Comment:
I don't think we need to discuss this in depth here given we're not going to do this in this PR but I think there's a race here - if you're incrementing the rule version on `getOrCreate` (while we hold a lock) there's a chance the version can be updated by another process and you end up operating on a stale version.

Reply:
That's not possible because the state is only updated by the scheduler which is a synchronous process. Everything else just reads this state.